tests/src/Unit/GetMatchingFieldTest.php

```diff
@@ -0,0 +1,130 @@
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Tests\UnitTestCase;
+use Drupal\og\OgGroupAudienceHelper;
+
+/**
+ * Tests the OgGroupAudienceHelper::getMatchingField method.
+ *
+ * @group og
+ *
+ * @coversDefaultClass \Drupal\og\OgGroupAudienceHelper
+ */
+class GetMatchingFieldTest extends UnitTestCase {
+
+
+  /**
+   * The entity type used for testing.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeInterface|\PHPUnit_Framework_MockObject_MockObject
+   */
+  protected $entityType;
+
+  /**
+   * The entity manager used for testing.
+   *
+   * @var \Drupal\Core\Entity\EntityManagerInterface|\PHPUnit_Framework_MockObject_MockObject
+   */
+  protected $entityManager;
+
+  /**
+   * The ID of the type of the entity under test.
+   *
+   * @var string
+   */
+  protected $entityTypeId;
+
+  /**
+   * The ID of the type of the bundle under test.
+   *
+   * @var string
+   */
+  protected $bundleId;
+
+  /**
+   * The group content entity.
+   *
+   * @var \Drupal\Core\Entity\ContentEntityInterface|\PHPUnit_Framework_MockObject_MockObject
+   */
+  protected $groupContent;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp() {
+    $field_name = 'test_field';
+
+    $this->entityTypeId = $this->randomMachineName();
+    $this->bundleId = $this->randomMachineName();
+    $this->provider = $this->randomMachineName();
+
+    $field_storage_definition_prophecy = $this->prophesize(FieldStorageDefinitionInterface::class);
+    $field_storage_definition_prophecy->getCardinality()
+      ->willReturn(FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED)
+      ->shouldBeCalled();
+
+    $field_definition_prophecy = $this->prophesize(FieldDefinitionInterface::class);
+    $field_definition_prophecy->getFieldStorageDefinition()
+      ->willReturn($field_storage_definition_prophecy->reveal())
+      ->shouldBeCalled();
+
+    $field_definition_prophecy->getType()
+      ->willReturn('og_membership_reference')
+      ->shouldBeCalled();
+
+    $field_definition_prophecy->getSetting('target_type')
+      ->willReturn($this->entityTypeId)
+      ->shouldBeCalled();
+
+
+    $this->groupContent = $this->prophesize(ContentEntityInterface::class);
+
+    $this->groupContent->getFieldDefinition($field_name)
+      ->willReturn($field_definition_prophecy->reveal());
+
+    $this->groupContent->bundle()
+      ->shouldBeCalled();
+    $this->groupContent->getEntityTypeId()
+      ->shouldBeCalled();
+
+    // If the cardinality is unlimited getting a count of the field items is
+    // never expected, so just check it's not called.
+    $this->groupContent->get($field_name)
+      ->shouldNotBeCalled();
+
+    $this->entityType = $this->getMock('\Drupal\Core\Entity\EntityTypeInterface');
+    $this->entityType->expects($this->any())
+      ->method('getProvider')
+      ->will($this->returnValue($this->provider));
+
+    $this->entityManager = $this->getMock('\Drupal\Core\Entity\EntityTypeManagerInterface');
```

Comment:
I agree, it's a bit strange to mix them up. Prophecy is easier to use. It does have its own limitations though.

Maybe try to prophesize it for now? If that doesn't solve it I'd be happy to dive in.